File: multi_agent/projects/im-mobile/lib/services/reminder_service.dart
// Message Reminder Service for IM Mobile
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/message_reminder.dart';

class ReminderService {
  final String baseUrl;
  final Map<String, String> _headers = {
    'Content-Type': 'application/json',
  };

  ReminderService({this.baseUrl = ''});

  void setUserId(int userId) {
    _headers['X-User-Id'] = userId.toString();
  }

  Future<MessageReminderModel> createReminder(ReminderRequest request) async {
    final response = await http.post(
      Uri.parse('$baseUrl/api/v1/reminders'),
      headers: _headers,
      body: jsonEncode(request.toJson()),
    );
    if (response.statusCode != 200) {
      final error = jsonDecode(response.body);
      throw Exception(error['message'] ?? 'Failed to create reminder');
    }
    return MessageReminderModel.fromJson(jsonDecode(response.body));
  }

  Future<MessageReminderModel> updateReminder(int id, ReminderRequest request) async {
    final response = await http.put(
      Uri.parse('$baseUrl/api/v1/reminders/$id'),
      headers: _headers,
      body: jsonEncode(request.toJson()),
    );
    if (response.statusCode != 200) {
      final error = jsonDecode(response.body);
      throw Exception(error['message'] ?? 'Failed to update reminder');
    }
    return MessageReminderModel.fromJson(jsonDecode(response.body));
  }

  Future<void> deleteReminder(int id) async {
    final response = await http.delete(
      Uri.parse('$baseUrl/api/v1/reminders/$id'),
      headers: _headers,
    );
    if (response.statusCode != 200) {
      final error = jsonDecode(response.body);
      throw Exception(error['message'] ?? 'Failed to delete reminder');
    }
  }

  Future<MessageReminderModel> dismissReminder(int id) async {
    final response = await http.post(
      Uri.parse('$baseUrl/api/v1/reminders/$id/dismiss'),
      headers: _headers,
    );
    if (response.statusCode != 200) {
      final error = jsonDecode(response.body);
      throw Exception(error['message'] ?? 'Failed to dismiss reminder');
    }
    return MessageReminderModel.fromJson(jsonDecode(response.body));
  }

  Future<List<MessageReminderModel>> getUserReminders() async {
    final response = await http.get(
      Uri.parse('$baseUrl/api/v1/reminders'),
      headers: _headers,
    );
    if (response.statusCode != 200) {
      final error = jsonDecode(response.body);
      throw Exception(error['message'] ?? 'Failed to get reminders');
    }
    final List<dynamic> data = jsonDecode(response.body);
    return data.map((e) => MessageReminderModel.fromJson(e)).toList();
  }

  Future<List<MessageReminderModel>> getPendingReminders() async {
    final response = await http.get(
      Uri.parse('$baseUrl/api/v1/reminders/pending'),
      headers: _headers,
    );
    if (response.statusCode != 200) {
      final error = jsonDecode(response.body);
      throw Exception(error['message'] ?? 'Failed to get pending reminders');
    }
    final List<dynamic> data = jsonDecode(response.body);
    return data.map((e) => MessageReminderModel.fromJson(e)).toList();
  }

  Future<int> getPendingCount() async {
    final response = await http.get(
      Uri.parse('$baseUrl/api/v1/reminders/count'),
      headers: _headers,
    );
    if (response.statusCode != 200) return 0;
    final data = jsonDecode(response.body);
    return data['count'] ?? 0;
  }

  DateTime setReminderTime(int minutesFromNow) {
    return DateTime.now().add(Duration(minutes: minutesFromNow));
  }

  String formatReminderDisplay(DateTime dateTime) {
    final now = DateTime.now();
    final diff = dateTime.difference(now);
    final minutes = diff.inMinutes;
    if (minutes < 60) return '$minutes min';
    final hours = diff.inHours;
    if (hours < 24) return '${hours}h';
    final days = diff.inDays;
    return '${days}d';
  }
}

File: im-desktop/src/services/reminder-service.ts
// Message Reminder Service for IM Desktop
import type { ReminderRequest, ReminderResponse } from '../types/message-reminder';

const API_BASE = '/api/v1/reminders';

export class ReminderService {
  private baseUrl: string;

  constructor(baseUrl: string = '') {
    this.baseUrl = baseUrl;
  }

  private getHeaders(): HeadersInit {
    const userId = localStorage.getItem('userId');
    return {
      'Content-Type': 'application/json',
      'X-User-Id': userId || '',
    };
  }

  async createReminder(request: ReminderRequest): Promise<ReminderResponse> {
    const response = await fetch(`${this.baseUrl}${API_BASE}`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify(request),
    });
    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to create reminder');
    }
    return response.json();
  }

  async updateReminder(id: number, request: ReminderRequest): Promise<ReminderResponse> {
    const response = await fetch(`${this.baseUrl}${API_BASE}/${id}`, {
      method: 'PUT',
      headers: this.getHeaders(),
      body: JSON.stringify(request),
    });
    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to update reminder');
    }
    return response.json();
  }

  async deleteReminder(id: number): Promise<void> {
    const response = await fetch(`${this.baseUrl}${API_BASE}/${id}`, {
      method: 'DELETE',
      headers: this.getHeaders(),
    });
    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to delete reminder');
    }
  }

  async dismissReminder(id: number): Promise<ReminderResponse> {
    const response = await fetch(`${this.baseUrl}${API_BASE}/${id}/dismiss`, {
      method: 'POST',
      headers: this.getHeaders(),
    });
    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to dismiss reminder');
    }
    return response.json();
  }

  async getUserReminders(): Promise<ReminderResponse[]> {
    const response = await fetch(`${this.baseUrl}${API_BASE}`, {
      method: 'GET',
      headers: this.getHeaders(),
    });
    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to get reminders');
    }
    return response.json();
  }

  async getPendingReminders(): Promise<ReminderResponse[]> {
    const response = await fetch(`${this.baseUrl}${API_BASE}/pending`, {
      method: 'GET',
      headers: this.getHeaders(),
    });
    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to get pending reminders');
    }
    return response.json();
  }

  async getPendingCount(): Promise<number> {
    const response = await fetch(`${this.baseUrl}${API_BASE}/count`, {
      method: 'GET',
      headers: this.getHeaders(),
    });
    if (!response.ok) return 0;
    const data = await response.json();
    return data.count || 0;
  }

  setReminderTime(messageId: number, minutesFromNow: number): string {
    const time = new Date(Date.now() + minutesFromNow * 60 * 1000);
    return time.toISOString();
  }

  parseReminderDateTime(dateStr: string): Date {
    return new Date(dateStr);
  }

  formatReminderDisplay(dateStr: string): string {
    const date = new Date(dateStr);
    const now = new Date();
    const diff = date.getTime() - now.getTime();
    const minutes = Math.floor(diff / 60000);
    if (minutes < 60) return `${minutes} min`;
    const hours = Math.floor(minutes / 60);
    if (hours < 24) return `${hours}h`;
    const days = Math.floor(hours / 24);
    return `${days}d`;
  }
}

export const reminderService = new ReminderService();
export default reminderService;
